flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/ReducingUpsertWriter.java

```diff
@@ -0,0 +1,179 @@
+    private final Map<RowData, Tuple2<RowData, Long>> reduceBuffer = new HashMap<>();
+    private final Function<RowData, RowData> keyExtractor;
+    private final Sink.ProcessingTimeService timeService;
+    private final long batchIntervalMs;
+
+    private boolean closed = false;
+    private int batchCount = 0;
+
+    ReducingUpsertWriter(
+            SinkWriter<RowData, ?, WriterState> wrappedWriter,
+            DataType physicalDataType,
+            int[] keyProjection,
+            SinkBufferFlushMode bufferFlushMode,
+            Sink.ProcessingTimeService timeService,
+            Function<RowData, RowData> valueCopyFunction) {
+        checkArgument(bufferFlushMode != null && bufferFlushMode.isEnabled());
+        this.wrappedWriter = checkNotNull(wrappedWriter);
+        this.timeService = checkNotNull(timeService);
+        registerFlush();
+        this.batchMaxRowNums = bufferFlushMode.getBatchSize();
+        this.batchIntervalMs = bufferFlushMode.getBatchIntervalMs();
+
+        List<LogicalType> fields = physicalDataType.getLogicalType().getChildren();
+        final RowData.FieldGetter[] keyFieldGetters =
+                Arrays.stream(keyProjection)
+                        .mapToObj(
+                                targetField ->
+                                        RowData.createFieldGetter(
+                                                fields.get(targetField), targetField))
+                        .toArray(RowData.FieldGetter[]::new);
+        this.keyExtractor = rowData -> createProjectedRow(rowData, RowKind.INSERT, keyFieldGetters);
+        this.valueCopyFunction = valueCopyFunction;
+    }
+
+    @Override
+    public void write(RowData element, Context context) throws IOException, InterruptedException {
+        wrappedContext.setContext(context);
+        addToBuffer(element, context.timestamp());
+    }
+
+    @Override
+    public List<Void> prepareCommit(boolean flush) throws IOException, InterruptedException {
+        flush();
+        return Collections.emptyList();
+    }
+
+    @Override
+    public List<WriterState> snapshotState() throws IOException {
+        return wrappedWriter.snapshotState();
+    }
+
+    @Override
+    public void close() throws Exception {
+        if (!closed) {
+            closed = true;
+            wrappedWriter.close();
+        }
+    }
+
+    private void addToBuffer(RowData row, Long timestamp) throws IOException, InterruptedException {
+        RowData key = keyExtractor.apply(row);
+        RowData value = valueCopyFunction.apply(row);
+        reduceBuffer.put(key, new Tuple2<>(changeFlag(value), timestamp));
+        batchCount++;
+
+        if (batchCount >= batchMaxRowNums) {
+            flush();
+        }
+    }
+
+    private void registerFlush() {
+        if (closed) {
+            return;
+        }
+        timeService.registerProcessingTimer(
+                System.currentTimeMillis() + batchIntervalMs, (t) -> flush());
```

Comment:
Currently, the timer is not reset on `flush`. Is this also the old behavior?

To reset the timer, you'd `registerFlush` on `flush` (which you forgot to do here as far as I can see) and also remember the the time at which it should trigger. Then you'd compare `t` against the time and ignore if it's smaller.

Reply:
> Currently, the timer is not reset on `flush`. Is this also the old behavior?

Yes, the timer was also not reset in the old implementation. I've adjusted the new version to the only flush if the last flush was before the flush interval.

> To reset the timer, you'd `registerFlush` on `flush` (which you forgot to do here as far as I can see) and also remember the the time at which it should trigger. Then you'd compare `t` against the time and ignore if it's smaller.

I now call `registerFlush` after every flushing in the processingTimer.

Reply:
Sounds good!